File: lib/data/models/appointment_model.dart
/// Production-grade Appointment Model
class Appointment {
  final String id;
  final String? patientId;
  final String? doctorId;
  final String doctorName;
  final String? doctorImage;
  final String specialty;
  final DateTime date;
  final String time;
  final String type; // 'video', 'clinic', 'home'
  final String status; // 'upcoming', 'completed', 'cancelled'
  final double fee;
  final String? notes;
  final DateTime? createdAt;

  Appointment({
    required this.id,
    this.patientId,
    this.doctorId,
    required this.doctorName,
    this.doctorImage,
    required this.specialty,
    required this.date,
    required this.time,
    required this.type,
    required this.status,
    required this.fee,
    this.notes,
    this.createdAt,
  });

  factory Appointment.fromJson(Map<String, dynamic> json) {
    return Appointment(
      id: json['id']?.toString() ?? '',
      patientId: json['patientId']?.toString(),
      doctorId: json['doctorId']?.toString(),
      doctorName: json['doctorName'] ?? '',
      doctorImage: json['doctorImage'],
      specialty: json['specialty'] ?? '',
      date: json['date'] is String
          ? DateTime.parse(json['date'])
          : json['date'] as DateTime,
      time: json['time'] ?? '',
      type: json['type'] ?? 'video',
      status: json['status'] ?? 'upcoming',
      fee: (json['fee'] ?? json['amount'] ?? 0).toDouble(),
      notes: json['notes'],
      createdAt: json['createdAt'] != null
          ? (json['createdAt'] is String
              ? DateTime.parse(json['createdAt'])
              : json['createdAt'] as DateTime)
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'patientId': patientId,
      'doctorId': doctorId,
      'doctorName': doctorName,
      'doctorImage': doctorImage,
      'specialty': specialty,
      'date': date.toIso8601String(),
      'time': time,
      'type': type,
      'status': status,
      'fee': fee,
      'notes': notes,
      'createdAt': createdAt?.toIso8601String(),
    };
  }

  Appointment copyWith({
    String? id,
    String? patientId,
    String? doctorId,
    String? doctorName,
    String? doctorImage,
    String? specialty,
    DateTime? date,
    String? time,
    String? type,
    String? status,
    double? fee,
    String? notes,
    DateTime? createdAt,
  }) {
    return Appointment(
      id: id ?? this.id,
      patientId: patientId ?? this.patientId,
      doctorId: doctorId ?? this.doctorId,
      doctorName: doctorName ?? this.doctorName,
      doctorImage: doctorImage ?? this.doctorImage,
      specialty: specialty ?? this.specialty,
      date: date ?? this.date,
      time: time ?? this.time,
      type: type ?? this.type,
      status: status ?? this.status,
      fee: fee ?? this.fee,
      notes: notes ?? this.notes,
      createdAt: createdAt ?? this.createdAt,
    );
  }

  bool get isUpcoming => status == 'upcoming' && date.isAfter(DateTime.now());
  bool get isCompleted => status == 'completed';
  bool get isCancelled => status == 'cancelled';
}

File: backend-new/src/modules/appointments/appointment.model.ts
import mongoose, { Schema, Document, Types } from 'mongoose';

export interface IAppointment extends Document {
  _id: Types.ObjectId;
  appointmentNumber: string;
  patientId: mongoose.Types.ObjectId;
  doctorId: mongoose.Types.ObjectId;
  hospitalId?: mongoose.Types.ObjectId;
  date: Date;
  time: string;
  type: 'video' | 'clinic' | 'home';
  status: 'pending' | 'confirmed' | 'completed' | 'cancelled' | 'rescheduled';
  reason?: string;
  symptoms?: string[];
  notes?: string;
  prescriptions?: string[];
  fee: number;
  paymentId?: mongoose.Types.ObjectId;
  paymentStatus: 'pending' | 'paid' | 'failed' | 'refunded';
  cancelReason?: string;
  cancelledBy?: 'patient' | 'doctor' | 'admin';
  cancelledAt?: Date;
  completedAt?: Date;
  reminderSent: boolean;
  createdAt: Date;
  updatedAt: Date;
}

const AppointmentSchema: Schema = new Schema(
  {
    appointmentNumber: {
      type: String,
      required: true,
      unique: true,
    },
    patientId: {
      type: Schema.Types.ObjectId,
      ref: 'User',
      required: [true, 'Patient ID is required'],
    },
    doctorId: {
      type: Schema.Types.ObjectId,
      ref: 'Doctor',
      required: [true, 'Doctor ID is required'],
    },
    hospitalId: {
      type: Schema.Types.ObjectId,
      ref: 'Hospital',
    },
    date: {
      type: Date,
      required: [true, 'Appointment date is required'],
    },
    time: {
      type: String,
      required: [true, 'Appointment time is required'],
    },
    type: {
      type: String,
      enum: ['video', 'clinic', 'home'],
      default: 'clinic',
    },
    status: {
      type: String,
      enum: ['pending', 'confirmed', 'completed', 'cancelled', 'rescheduled'],
      default: 'pending',
    },
    reason: {
      type: String,
    },
    symptoms: [{
      type: String,
    }],
    notes: {
      type: String,
    },
    prescriptions: [{
      type: String,
    }],
    fee: {
      type: Number,
      required: [true, 'Appointment fee is required'],
      min: 0,
    },
    paymentId: {
      type: Schema.Types.ObjectId,
      ref: 'Payment',
    },
    paymentStatus: {
      type: String,
      enum: ['pending', 'paid', 'failed', 'refunded'],
      default: 'pending',
    },
    cancelReason: {
      type: String,
    },
    cancelledBy: {
      type: String,
      enum: ['patient', 'doctor', 'admin'],
    },
    cancelledAt: {
      type: Date,
    },
    completedAt: {
      type: Date,
    },
    reminderSent: {
      type: Boolean,
      default: false,
    },
  },
  {
    timestamps: true,
  }
);

// Generate appointment number
AppointmentSchema.pre('save', async function (next) {
  if (!this.appointmentNumber) {
    const count = await mongoose.model('Appointment').countDocuments();
    this.appointmentNumber = `APT${String(count + 1).padStart(6, '0')}`;
  }
  next();
});

// Indexes
AppointmentSchema.index({ patientId: 1, date: -1 });
AppointmentSchema.index({ doctorId: 1, date: 1 });
AppointmentSchema.index({ status: 1, date: 1 });

export const Appointment = mongoose.model<IAppointment>('Appointment', AppointmentSchema);
